### PR TITLE
Promisify fromXML API

### PR DESCRIPTION
**Pull request by oguzeroglu**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -47,32 +47,36 @@ var reader = new Reader(model);
...
-    //        { $type: 'my:Engine', powser: 121, fuelConsumption: 10 }
-    //      ]
-    //    }
-    //  ]
-    // }
+try {
+  var result = await reader.fromXML(xml, rootHandler);
+
+  const {
+    element,
```
Comment (nikku) [inline README.md:54]:
```suggestion
    element: cars,
```

File: README.md [hunk 2]
```diff
@@ -47,32 +47,36 @@ var reader = new Reader(model);
...
+    parseContext
+  } = result;
+
+  if (parseContext.warnings.length) {
+    console.log('import warnings', parseContext.warnings);
   }
-});
+
+  console.log(element); // cars
```
Comment (nikku) [inline README.md:62]:
```suggestion
  console.log(cars);
```

File: lib/read.js [hunk 3]
```diff
@@ -842,11 +840,9 @@ Reader.prototype.fromXML = function(xml, options, done) {
     .on('error', handleError)
     .on('warn', handleWarning);
 
-  // deferred parse XML to make loading really ascnchronous
```
Comment (nikku) [inline lib/read.js:845]: Can we keep this comment?

File: lib/read.js [hunk 4]
```diff
@@ -859,13 +855,19 @@ Reader.prototype.fromXML = function(xml, options, done) {
...
-    // handle the situation that we could not extract
-    // the desired root element from the document
     if (!err && !element) {
       err = error('failed to parse document as <' + rootHandler.type.$descriptor.name + '>');
     }
 
-    done(err, err ? undefined : element, context);
+    if (err) {
+      err.parseContext = context;
+      reject(err);
```
Comment (nikku) [inline lib/read.js:864]: Please do `return` along with resolving / rejecting:

```suggestion
      return reject(err);
```

File: lib/read.js [hunk 5]
```diff
@@ -859,13 +855,19 @@ Reader.prototype.fromXML = function(xml, options, done) {
...
+    } else {
+      resolve({
```
Comment (nikku) [inline lib/read.js:866]:
```suggestion
      return resolve({
```

File: test/spec/reader.js [hunk 6]
```diff
@@ -14,6 +14,16 @@ describe('Reader', function() {
 
   var createModel = createModelBuilder('test/fixtures/model/');
 
+  var read = function(reader, xml, optionsOrRootHandler, callback) {
```
Comment (nikku) [inline test/spec/reader.js:17]: Please transform into awaitable method (drop callback).

File: test/spec/reader.js [hunk 7]
```diff
@@ -28,7 +38,7 @@ describe('Reader', function() {
       var xml = '<props:complexAttrs xmlns:props="http://properties"></props:complexAttrs>';
 
       // when
-      reader.fromXML(xml, rootHandler, function(err, result, context) {
+      read(reader, xml, rootHandler, function(err, result, context) {
```
Comment (nikku) [inline test/spec/reader.js:41]: Please transform the tests into `async/await` style ones:


```
it('should do stuff', async function() {
  
   // when
   var {
      element,
      parseContext
   } = await reader.fromXML(...);

   ...
});
```

Comment (oguzeroglu) [reply test/spec/reader.js]: regex FTW

**Comment by oguzeroglu** [comment]
IMO this is ready to be merged now.

**Comment by nikku** [comment]
Will look into it.

**Comment by nikku** [comment]
I've pushed two commits, call them review hints, on top: 

* https://github.com/bpmn-io/moddle-xml/pull/45/commits/21704516a56bf5750b6e6f83f5485f48589ac941 :arrow_right: simplifies the awaitable result.
* https://github.com/bpmn-io/moddle-xml/pull/45/commits/2f71fd2249507af39ba2234419009f46e274ab76 :arrow_right: update docs accordingly

Please double check and :+1: if you agree.

**Comment by oguzeroglu** [comment]
I like it 👍

**Comment by nikku** [comment]
Squashed and merged via https://github.com/bpmn-io/moddle-xml/commit/39d2507aa746df4c0c37ba42025195beb7395b1d.

**Comment by nikku** [comment]
Tagged as BREAKING CHANGE.

**Pull request merged**